erts/emulator/sys/unix/erl_child_setup.c

```diff
@@ -572,97 +581,103 @@ main(int argc, char *argv[])
                 if (isatty(0) && isatty(1)) {
                     tcsetattr(0,TCSANOW,&initial_tty_mode);
                 }
+                kill_all_children();
                 _exit(0);
             }
             /* Since we use unix domain sockets and send the entire data in
                one go we *should* get the entire payload at once. */
             ASSERT(res == sizeof(proto));
-            ASSERT(proto.action == ErtsSysForkerProtoAction_Start);
+            if (proto.action == ErtsSysForkerProtoAction_Start) {
+                sys_sigblock(SIGCHLD);
 
-            sys_sigblock(SIGCHLD);
+                errno = 0;
 
-            errno = 0;
-
-            os_pid = fork();
-            if (os_pid == 0)
-                start_new_child(pipes);
+                os_pid = fork();
+                if (os_pid == 0) {
+                    start_new_child(pipes);
+                }
 
-            add_os_pid_to_port_id_mapping(proto.u.start.port_id, os_pid);
+                {
+                    ErtsSysExitStatus es;
+                    es.os_pid = os_pid;
+                    es.port_id = proto.u.start.port_id;
+                    es.want_exit_status = proto.u.start.want_exit_status;
+                    hash_put(forker_hash, &es);
+                }
 
-            /* We write an ack here, but expect the reply on
-               the pipes[0] inside the fork */
-            proto.action = ErtsSysForkerProtoAction_Go;
-            proto.u.go.os_pid = os_pid;
-            proto.u.go.error_number = errno;
-            write_all(pipes[1], (char *)&proto, sizeof(proto));
+                /* We write an ack here, but expect the reply on
+                the pipes[0] inside the fork */
+                proto.action = ErtsSysForkerProtoAction_Go;
+                proto.u.go.os_pid = os_pid;
+                proto.u.go.error_number = errno;
+                write_all(pipes[1], (char *)&proto, sizeof(proto));
 
 #ifdef FORKER_PROTO_START_ACK
-            proto.action = ErtsSysForkerProtoAction_StartAck;
-            write_all(uds_fd, (char *)&proto, sizeof(proto));
+                proto.action = ErtsSysForkerProtoAction_StartAck;
+                write_all(uds_fd, (char *)&proto, sizeof(proto));
 #endif
 
-            sys_sigrelease(SIGCHLD);
-            close(pipes[0]);
-            close(pipes[1]);
-            close(pipes[2]);
+                sys_sigrelease(SIGCHLD);
+                close(pipes[0]);
+                close(pipes[1]);
+                close(pipes[2]);
+            } else if (proto.action == ErtsSysForkerProtoAction_Stop) {
+                ErtsSysExitStatus est, *es;
+                est.os_pid = proto.u.stop.os_pid;
+                es = hash_remove(forker_hash, &est);
```

Comment:
Why do we remove it from the hash here? Do we not want closed but alive processes to be killed when the VM shuts down?

Reply:
We no longer want to receive an exit message from the spawned process, and I was getting conservative about memory, but I now see that we can keep both behaviors by changing the hash_remove to:
`es.want_exit_status = false`

(Also, my concern about memory is nonsense since the handful of bytes in `es` is tracking an entire, in-memory child process.)